---
title: "qtl_analysis"
author: "SERGIO PEREZ-LIMON"
date: "3/7/2021"
output: 
  html_document: 
    keep_md: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries, set seed and set working directory

```{r}
library(tidyverse)
library(googlesheets4)
library(googledrive)
library(qtl)
library(broom)
library(pals)


# Set the seed as the same value as previous analysis
set.seed(54955149)

# Note: don't forget to set working directory the same as project directory
```

Functions

```{r}

# transform p.value to significance codes
p.value_to_ast <- function(p.value) {
  
  ast <- case_when(
    p.value < 0.001 ~ "***",
    p.value >= 0.001 & p.value < 0.01 ~ "**",
    p.value >= 0.01 & p.value < 0.05 ~ "*",
    p.value >= 0.05 & p.value < 0.01 ~ ".",
    p.value > 0.01 ~ "NS",
    T ~ NA_character_)
  
  return(ast)
  
}

# raincloud plots function
raincloud_plot <- function(data, tissue) {
  
  plot <- data %>%
    ggplot(data =., 
           aes(x = hun, 
               y = value, 
               fill = hun)
           ) +
    ggdist::stat_halfeye(aes(fill = hun),
                         adjust = .5, 
                         width = .6, 
                         .width = 0, 
                         justification = -.2, 
                         point_colour = NA,
                         color = "black",
                         ) +
    geom_boxplot(
      width = .15, 
      outlier.shape = NA,
      color = "black"
      ) +
    gghalves::geom_half_point(
      ## draw jitter on the left
      side = "l", 
      ## control range of jitter
      range_scale = .4, 
      ## add some transparency
      alpha = .5,
      shape = 21,
      color = "black"
      ) +
    ggpubr::stat_compare_means(
      comparisons = list(c("HUN", "WT")),
      label = "p.signif",
      method = "wilcox.test") +
    xlab(NULL) +
    ylab("ppm") +
    theme(
      panel.background = element_rect(fill = "white", 
                                      colour ="grey50"),
      plot.title = element_text(hjust = 0.5)
      ) +
    scale_fill_manual(
      values = hun_color_scheme, guide = NULL
      ) +
    ggtitle(tissue)
  
  return(plot)
}

# transform summary and add information in a suitable tibble
summary.to.tibble <- function(CROSS, SCANONE, PERMS, ...) {
  
  a <- summary(
    object = SCANONE, 
    perms = PERMS,
    format = "tabByChr",
    ci.function = "lodint",
    ...
  ) 
  
  
  b <- summary(PERMS, ...)
  
  c <- tibble(
    phenotype = b %>% attr("dimnames") %>% .[[2]],
    threshold = b %>% as.numeric() %>% round(., 3)
  )
  
  p1 <- a %>%
    map_df(
      .x = ., 
      .f = ~ as_tibble(.x, rownames = "qtl"), 
      .id = "chr"
    ) %>%
    separate(
      qtl, 
      into = c("phenotype", "marker"), 
      sep = " : "
    ) %>%
    arrange(chr, pos, phenotype) %>%
    mutate(
      marker = map2_chr(
        .x = chr,
        .y = pos,
        .f = ~ find.marker(cross = CROSS,
                           chr = .x,
                           pos = .y)))
  
  p2 <- p1 %>%
    pivot_longer(c(pos, ci.low, ci.high)) %>%
    mutate(
      c_marker = map2_chr(
        .x = chr,
        .y = value,
        .f = ~ find.marker(
          cross = CROSS,       
          chr = .x,
          pos = .y
        )
      )) %>%
    mutate(c_marker = gsub("^\\d{1,}_", "", c_marker) %>% as.integer,
           c_marker = round(c_marker/1e6, 3),
           name = paste0("p_", name)) %>%
    select(-value) %>%
    pivot_wider(names_from = name, values_from = c_marker) %>%
    mutate(interval = p_ci.high - p_ci.low)
  
  p3 <- p1 %>%
    left_join(p2, by = c("phenotype", "marker", "chr", "lod")) %>%
    left_join(c, by = "phenotype")
  
  
  return(p3)
}

# transform results in df to use to plot in ggplot2
lodplot_map_df <- function(SCANONE, PERMS, SUMMARY, ...) {
  
  a <- SCANONE %>%
    as_tibble(rownames = "marker")
  
  dist_cM <- a %>%
    select(chr, pos) %>%
    group_by(chr) %>%
    filter(pos == max(pos)) %>%
    ungroup %>%
    mutate(fin = lag(pos),
           space = (5),
           pen = ifelse(is.na(fin), 0, pos + fin + space),
           pen2 = cumsum(pen)) %>%
    select(chr, pen2)
  
  b <- summary(PERMS, ...)
  
  c <- tibble(
    phenotype = b %>% attr("dimnames") %>% .[[2]],
    threshold = b %>% as.numeric() %>% round(., 3)
  )
  
  d <- SUMMARY %>%
    select(phenotype, chr, ci.low, ci.high)
  
  e <- a %>%
    left_join(dist_cM, by = "chr") %>%
    mutate(cM = pos + pen2) %>%
    select(-pen2) %>%
    select(marker:pos, cM, names(.)) %>%
    pivot_longer(
      cols = -c(marker:cM), 
      names_to = "phenotype", 
      values_to = "lod"
    ) %>%
    left_join(c, by = c("phenotype")) %>%
    left_join(d, by = c("chr", "phenotype")) %>%
    mutate(y = pmap_dbl(
      .l = .,
      .f = ~ with( 
        list(...),
        ifelse(dplyr::between(pos, ci.low, ci.high),
               lod,
               NA_real_
               ))))  %>%
    mutate(alpha = ifelse(is.na(y), 0, 1)) %>%
    mutate(chr = as.integer(chr))
  
  breaks <- e %>%
    group_by(chr) %>%
    summarise(breaks = mean(cM))
  
  f <- e %>%
    left_join(breaks, by = "chr")
  
  return(f)
}

# transform MQM results in df to use to plot in ggplot2
mapa_base_MQM <- function(SCANONE, TIBBLE) {
  
  escan <- eval(SCANONE) %>% 
    as_tibble(., rownames = "marker") %>% 
    rename(pos = `pos (cM)`) %>%
    select(-info, - `LOD*info`) %>%
    rename_at(.vars = vars(contains("LOD")), 
              .funs = list(~ gsub("LOD ", "", .)))
  
  
  summ <- TIBBLE %>%
    select(trait, chr, ci.low, ci.high, treshold) %>%
    rename(chr_sum = chr)
  
  mb <- escan %>%
    select(1:3) %>%
    mutate(chr = as.double(chr)) %>%
    group_by(chr) %>%
    nest() %>%
    mutate(max = purrr::map_dbl(.x = data, .f = ~ max(.x$pos))) %>%
    ungroup() %>%
    mutate(comp = lag(cumsum(max)),
           comp2 = (as.numeric(chr)-1)* 25) %>%
    mutate(comp = ifelse(is.na(comp), 0, comp)) %>%
    unnest(data) %>%
    mutate(cM = pos + comp + comp2) %>%
    select(-(max:comp2)) %>%
    group_by(chr) %>% 
    mutate(max = max(cM), 
           min = min(cM), 
           breaks = mean(cM), 
           limit_low = min(cM),
           limit_high= max(cM), 
           inicio = head(cM, n = 1),
           fin = tail(cM, n = 1)) %>% 
    select(marker, chr, cM, pos, (breaks:fin)) %>%
    left_join(., escan %>% select(-chr, -pos)) %>%
    pivot_longer(names_to = "trait", values_to = "lod", cols = -(marker:fin)) %>%
    ungroup()
  
  mb_2 <- mb %>%
    left_join(., summ, by = c("trait" = "trait", "chr" = "chr_sum")) %>%
    mutate(alpha = pmap_dbl(.l = .,
                            .f = ~ case_when(is.na(..12) | is.na(..13) ~ 0,
                                             between(..4, ..12, ..13) ~ 1,
                                             T ~ 0)))
  return(mb_2)
}

MQM_summary_to_tibble <- function(MQM_perms_df, CROSS, ...) {
  
  mqm_summ <- MQM_perms_df %>% 
    mutate(sum = map2(.x = mqmscan, 
                      .y = perms_2,
                      .f = ~ summary(.x, 
                                     perms = .y, 
                                     alpha = 0.05, 
                                     format = "tabByChr")))
  summary <- mqm_summ %>%
  select(sum, perms_2) %>%
  mutate(treshold = purrr::map_dbl(.x = perms_2, 
                                   .f = ~ summary(object = .x, alpha = 0.05) %>%
                                     as.double(.) %>% 
                                     round(., 2))) %>%
  unnest(sum) %>% 
  filter(!map_lgl(.x = sum, .f = ~ is.null(.x))) %>%  
  mutate(tib = purrr::map(.x =sum, .f = ~ as_tibble(.x, rownames = "trait"))) %>%
  unnest(tib) %>% 
  select(-sum, -perms_2) %>%
  separate(trait, into = c("trait", "marker"), sep = " : ") %>%
  mutate(lodtype = ifelse(grepl("LOD\\*info", trait), "lod*info", "lod")) %>%
  pivot_wider(names_from = lodtype, values_from = lod) %>%
  mutate(`lod*info` = ifelse(is.na(`lod*info`), lead(`lod*info`), `lod*info`)) %>%
  filter(! grepl("LOD\\*info", trait)) %>%
  rename(pos = `pos (cM)`) %>%
  mutate(trait = gsub("LOD ", "", trait),
         marker = pmap_chr(.l = ., 
                           .f = ~ ifelse(!grepl("^(\\d+)\\_", ..3),
                                         find.marker(cross = CROSS, 
                                                     chr = ..4, 
                                                     pos = ..5),
                                         ..2)),
         p_pos = round(as.double(gsub("^(\\d+)\\_", "", marker))/1e6, 2),
         p_pos_l = map2_dbl(.x = chr, .y = ci.low, 
                            .f = ~ find.marker(cross = CROSS,
                                               chr = .x,
                                               pos = .y) %>%
                              gsub("^(\\d+)\\_", "", .) %>% 
                              as.double(.)/1e6) %>% 
           round(.,2),
         p_pos_h = map2_dbl(.x = chr, 
                            .y = ci.high, 
                            .f = ~ find.marker(cross = CROSS,
                                               chr = .x, 
                                               pos = .y) %>%
                              gsub("^(\\d+)\\_", "", .) %>% 
                              as.double(.)/1e6) %>% 
           round(.,2),
         interval = p_pos_h - p_pos_l) %>%
    mutate(chr = as.double(chr)) %>%
    arrange(chr) %>%
    mutate(method = "MQM")
  return(summary)
}



```
Importing data

Data that is goint to be used:

1. Ionomics Data
2. CML312 x W22 HUN mapping population genetic map

```{r}
phenotypes_url <- "https://docs.google.com/spreadsheets/d/1zC8c44uLXKEccDdUKaMmhRGn0JU5wNt1iOYS1xmPWGs/edit#gid=1931442445"

gen_map_url <- "https://docs.google.com/spreadsheets/d/1Abx_s6XxjG91YlLIaFxNn4HNcbnA3lKRgR0jw2DZZek/edit#gid=1187220000"

# Importing lsmeans from the sheets document
ionomics_phenotypes_raw <- phenotypes_url %>%
  as_id %>%
  range_read(., na = "NA") %>%
  select(id = GEN, hun = HUN, matches("seed|leaf"))

#Importing the genetic map of the CML312xW22 HUN F2 mapping population
gen_map <- gen_map_url %>%
  as_id %>%
  range_read(., na = "NA", col_types = "c") 

gen_map <- gen_map %>%
  mutate(id = ifelse(is.na(id), "", id))
```

Transforming data:

- Selecting ionomic phenotypes for leaf and seed
- Declaring color scheme for HUN levels: hun (Non-Pikachu yellow) and wt (green).

```{r}
ionomics_phenotypes <- ionomics_phenotypes_raw %>%
  filter(id != "ARES") %>%
  pivot_longer(-c(id, hun)) %>%
  mutate(name = gsub("\\.SR", "", name)) %>%
  separate(name, into = c("ion", "tissue")) %>%
  filter(!is.na(value))

# Declaring color scheme for hun 
hun_color_scheme <- c("HUN" = "#EDB120", "WT" = "#77AC30")
```

Declaring input and output directories in drive

```{r}
output_tables_drive_id <- "https://drive.google.com/drive/folders/1xYdj_AHkOsQaOoY6SgKDTuHZBoiww9Ci" %>%
  as_id

output_plots_drive_id <- "https://drive.google.com/drive/folders/1L1gh2-bIE_0CGPVt7iy7MVL7YuIs0xDD" %>%
  as_id

input_data_drive_id <-  "https://drive.google.com/drive/u/0/folders/19bOcGgnv6YvjdoQtVXsehf6_2NpCkynS" %>%
  as_id
```

First Look at data.

1. Compare means of HUN and WT families using Kruskal-Wallis non parametric test for every ion and tissue.
2. Make raincloud-plots by ion and tissue, and compare between medians using Kruskal-Wallis non parametric test.

```{r}
# create directory for results
# dir.create(file.path(getwd(), "output"))
# dir.create(file.path(paste0(getwd(), "/output"), "tables"))

# Compare distributions with kruskal-wallis test
ion_tissue_kw <- ionomics_phenotypes %>%
  group_by(ion, tissue) %>%
  nest %>%
  mutate(kw = map(.x = data, 
                  .f = ~ kruskal.test(value ~ hun, 
                                      data = .x))) %>%
  mutate(kw_tidy = map(.x = kw, .f = ~ tidy(.x))) %>%
  ungroup %>%
  select(ion, tissue, kw_tidy) %>%
  unnest(kw_tidy) %>%
  select(ion:p.value) %>%
  mutate(sig = p.value_to_ast(p.value)) %>%
  arrange(ion, tissue)

# save kruskal-wallis test df
write_csv(ion_tissue_kw,
          "output/tables/ion_tissue_kruskal_wallis_test.csv")

# # upload table to drive
# drive_upload(
#   media = "output/tables/ion_tissue_kruskal_wallis_test.csv",
#   path = output_tables_drive_id,
#   name = "ion_tissue_kruskal_wallis_test.csv",
#   type = "spreadsheet"
#   )

# Note: comparison of distributions using kruskal-wallis
ionomics_raincloud_plots <- ionomics_phenotypes %>%
  group_by(ion, tissue) %>%
  mutate(hun = factor(hun, levels = c("WT", "HUN"))) %>%
  nest  %>%
  mutate(plot = map2(.x = data, 
                     .y = tissue,
                     .f = ~ raincloud_plot(data = .x, 
                                           tissue = .y))) %>%
  select(ion, plot) %>%
  group_by(ion) %>%
  summarise(plot = list(plot)) %>% 
  mutate(title = paste0("Ion: ", ion)) %>%
  mutate(wrap = map2(
    .x = plot, 
    .y = title,
    .f = ~ patchwork::wrap_plots(.x, guides = "collect") +
      patchwork::plot_annotation( 
        title = .y,
        theme = theme(plot.title = element_text(hjust = 0.5)
                      )
        )
    )) %>%
  ungroup

p <- ionomics_raincloud_plots$wrap

# # create directory for plots
# dir.create(file.path(paste0(getwd(), "/output"), "plots"))
# dir.create(file.path(paste0(getwd(), "/output/plots"), "rainclowd_plots"))

# Saving plots into a single pdf
pdf("output/plots/rainclowd_plots/ions_raincloud_plots.pdf",
    onefile = TRUE)
map(p, ~ print(.x))
dev.off()

# #save raincloud plots to google drive
# drive_upload(
#   media = "output/plots/rainclowd_plots/ions_raincloud_plots.pdf",
#   path = output_plots_drive_id,
#   name = "ions_raincloud_plots.pdf",
#   type = "pdf"
#   )

```

Create CrossObject

1. Create new traits: proportion between leaf/seed values
2. Create CrossObject

```{r}
# ion phenotypes
phenotypes_ion <- ionomics_phenotypes %>%
  mutate(tissue = tolower(tissue)) %>%
  mutate(name = paste0(ion, "_", tissue)) %>%
  select(-c(ion, tissue)) %>%
  pivot_wider(names_from = name, values_from = value)

# ratio between leaf/seed
phenotypes_ratio <- ionomics_phenotypes %>%
  mutate(tissue = tolower(tissue)) %>%
  pivot_wider(names_fro = tissue, values_from = value) %>%
  mutate(ratio = leaf/seed) %>%
  select(-c(leaf:seed)) %>%
  mutate(ion = paste0(ion, "_ratio")) %>%
  pivot_wider(names_from = ion, values_from = ratio)

# mean between leaf/seed
phenotypes_mean <- ionomics_phenotypes %>%
  mutate(tissue = tolower(tissue)) %>%
  pivot_wider(names_from = tissue, values_from = value) %>%
  rowwise() %>%
  mutate(mn = mean(c(leaf, seed))) %>%
  select(-c(leaf:seed)) %>%
  mutate(ion = paste0(ion, "_mean")) %>%
  pivot_wider(names_from = ion, values_from = mn)

# join all phenotypes
phenotypes_all <- left_join(phenotypes_ion,
                            phenotypes_ratio) %>%
  left_join(phenotypes_mean) %>%
  select(id, hun, sort(names(.)))

# join all phenotypes with the genetic map and create the CrossObject
crossobject_ion_hun <- gen_map %>%
  mutate(enchar = nchar(id)) %>%
  mutate(id = ifelse(enchar != 6, 
                     gsub("HUN", "HUN0", id),
                     id)) %>%
  select(-enchar) %>%
  filter(id %in% c("", phenotypes_all$id)) %>%
  left_join(phenotypes_all) %>%
  select(all_of(names(phenotypes_all)), names(.)) %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(all_of(names(phenotypes_all)),
                ~ ifelse(is.na(.) & row_number() < 3, "", .)))

# # create directory for the CrossObject
# dir.create(file.path(paste0(getwd(), "/output"), "crossobject"))

# save crossobject
write_csv(crossobject_ion_hun,
          "output/crossobject/crossobject_ion_hun.csv")

# # upload crossobject to drive
# drive_upload(
#   media = "output/crossobject/crossobject_ion_hun.csv",
#   path = input_data_drive_id,
#   name = "crossobject_ion_hun.csv",
#   type = "spreadsheet"
#   )
```

QTL analysis

1. Import CrossObject
2. Create a pheno.col df
3. Identify covariate (genotype on marker 5_2269274)
4. Calculate genoprob and simprob
5. Scanone with an additive covariate model 

```{r}
# Import CrossObject
ion_crossobject <-  read.cross(
  format = "csv",
  dir = "output/crossobject",
  file = "crossobject_ion_hun.csv",
  genotypes = c("AA", "AB", "BB"),
  alleles = c("CML", "W22"),
  BC.gen = 0,
  F.gen = 2,
  estimate.map = F)

# creating a pheno.col dataframe
ion_pheno.col <- ion_crossobject %>%
  phenames %>%
  enframe(name = "pheno.col", value = "phenotype") %>%
  filter(row_number() > 2) 
  
# Identifying covariate (hun genotype)
hun_cov <- pull.geno(ion_crossobject, chr = 5) %>% .[,"5_2269274"]
hun_cov[hun_cov == 3] <- 0

#NOTE: WT = 1; HUN = 0.

# Create directory to save the results

# dir.create(file.path(paste0(getwd(), "/output"), "qtl_objects"))
# dir.create(file.path(paste0(getwd(), "/output/qtl_objects"), "scanone"))
# dir.create(file.path(paste0(getwd(), "/output/qtl_objects"),
#                      "permutations"))
# dir.create(file.path(paste0(getwd(), "/output/tables"),
#                      "scanone_summary"))

# calculate the genetic probabilities
ion_crossobject <- calc.genoprob(
  ion_crossobject, 
  step = 0.5
  )

# simulating genotypes
ion_crossobject <- sim.geno(
  ion_crossobject,
  n.draws = 128, 
  step = 0.5, 
  error.prob = 0.001
  )
```

QTL analysis using MQM 

```{r}

MQM_ion_crossobject <-  read.cross(
  format = "csv",
  dir = "output/crossobject",
  file = "crossobject_ion_hun.csv",
  genotypes = c("AA", "AB", "BB"),
  alleles = c("CML", "W22"),
  estimate.map = F)

# creating a pheno.col dataframe
MQM_ion_pheno.col <- MQM_ion_crossobject %>%
  phenames %>%
  enframe(name = "pheno.col", value = "phenotype") %>%
  filter(row_number() > 2) 

# genotype imputation (augmentation)
MQM_ion_hun_add <- mqmaugment(MQM_ion_hun_add, minprob = 0.1)

# Set automatic cofactors across all genome (100)
MQM_ion_hun_add_cofactors <- mqmautocofactors(MQM_ion_hun, 100)


# Multiqtl scan using MQM
MQM_ion_hun_add_scan <- MQM_ion_pheno.col %>%
  mutate(
    mqmscan = 
      map(.x = pheno.col,
          .f = ~ mqmscan(
            cross = MQM_ion_hun_add, 
            cofactors = MQM_ion_hun_add_cofactors,
            pheno.col = .x, 
            step.size = 1,
            window.size = 25,
            n.clusters = 4)))

# Permutation test (1000 perms)
MQM_ion_hun_add_scan_perms <- MQM_RIL_scan_perms

MQM_ion_hun_add_scan_perms <- MQM_ion_hun_add_scan %>%
  mutate(perms = purrr::map(
    .x = pheno.col,
    .f = ~ mqmpermutation(
      cross = MQM_ion_hun_add, 
      pheno.col = .x, 
      method = "permutation",
      n.cluster = 4,
      n.perm = 1000,
      verbose = F
      )))

# save scanone
save(
  list = c("ion_crossobject_intcov.out.hk"), 
  file = "output/qtl_objects/scanone/MQM_ion_hun_add_scan_perms.RData"
)

load("output/qtl_objects/scanone/MQM_ion_hun_add_scan_perms.RData")

MQM_ion_hun_add_scan_perms2 <- MQM_ion_hun_add_scan_perms %>%
  mutate(perms_2 = purrr::map(.x = perms, .f = ~ mqmprocesspermutation(.x)))


MQM_ion_hun_summary <- MQM_ion_hun_add_scan_perms2 %>%
  MQM_summary_to_tibble(., CROSS = MQM_ion_hun_add) %>%
  mutate(chr = as.character(chr)) %>%
  mutate(marker = pmap_chr(
    .l =.,
    .f = ~ with(
      list(...),
      find.marker(
        cross = ion_crossobject,
        chr = chr, 
        pos = pos
      )))) %>%
  mutate(p_pos = gsub("^\\d{1,2}_", "", marker) %>% as.integer,
         p_pos = p_pos/1e6,
         p_pos_l = ifelse(ci.low == 0, 0, p_pos_l)) %>%
  select(-c(`lod*info`, method)) %>%
  rename(phenotype = trait, p_ci.low = p_pos_l, p_ci.high = p_pos_h,
         threshold = treshold)

MQM_ion_hun_add_LOD_map <- MQM_RIL_scan_perms_2 %>%
  mutate(
    mapa = map(
      .x = mqmscan,
      .f = ~ mapa_base_MQM(
        SCANONE = .x, 
        TIBBLE = MQM_RIL_QTL_summary))) %>%
  .$mapa %>%
  Reduce(bind_rows, .)

MQM_lodplot_function <- function(data_df, ion) {
  
  title <- paste0("Ion: ", ion)
  
  plot <- data_df %>% 
    {
      ggplot(data =., 
             aes(x = cM, 
                 y = lod, 
                 group = interaction(chr), 
                 alpha = alpha)) +
        geom_line(size = 1) +
        geom_line(aes(x = cM, 
                      y = lod, 
                      group = (chr)), 
                  alpha = 0.15, 
                  size = 1.5) +
        geom_vline(xintercept = c(.$inicio, .$fin), 
                   linetype = "dotted") +
        geom_hline(aes(yintercept = treshold),
                   linetype = "dotted", 
                   size = 0.75, 
                   alpha = 0.75) +
        theme(panel.background = 
                element_rect(
                  fill = "white", 
                  colour = "grey50")) +
        scale_x_continuous(
          name = "chr", 
          breaks = unique(.$breaks), 
          labels = unique(as.character((.$chr)))) +
        scale_alpha_identity() + 
        theme(legend.position = "top",
              text = element_text(size = 12), 
              plot.title = element_text(hjust = 0.5)) +
        xlab(NULL) +
        ylab("LOD") +
        ggtitle(title) +
        facet_grid(trait ~ .) 
        
    }
  
  return(plot)
}

MQM_ion_hun_add_lodplots <- MQM_ion_hun_add_LOD_map %>%
  group_by(trait) %>%
  nest %>%
  separate(trait, into = c("ion", NA), remove = F) %>%
  unnest(data) %>%
  group_by(ion) %>%
  nest %>%
  mutate(plot = map2(.x = data,
                     .y = ion, 
                     .f = ~ MQM_lodplot_function(.x, .y))) 
  
p <- MQM_ion_hun_add_lodplots$plot

# create directory for plots
# dir.create(file.path(paste0(getwd(), "/output/plots"), "MQM_lodplots"))

# Saving plots into a single pdf
ggsave(
  filename = 
    "output/plots/MQM_lodplots/ion_MQM_lodplots.pdf", 
  plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
  width = 15, height = 9
)

```

QTL analysis with an additive covariate model

```{r}
######################################################
# Using scanone applying an additive covariate model #
######################################################

ion_crossobject_addcov.out.hk <- 
  scanone(ion_crossobject, 
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          addcovar = hun_cov
          )

# save scanone
save(
  list = c("ion_crossobject_addcov.out.hk"), 
  file = "output/qtl_objects/scanone/ion_crossobject_addcov.out.hk.RData"
)

# load("output/qtl_objects/scanone/ion_crossobject_addcov.out.hk.RData")

# Permutation test (1000 permutations)
ion_crossobject_addcov.operm <- 
  scanone(ion_crossobject, 
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          addcovar = hun_cov,
          n.perm = 1000
          )
# save permutation test
save(
  list = c("ion_crossobject_addcov.operm"), 
  file = "output/qtl_objects/permutations/ion_crossobject_addcov.operm.RData"
)

load("output/qtl_objects/permutations/ion_crossobject_addcov.operm.RData")

# Summary, alpha = 0.05
summary(
  ion_crossobject_addcov.out.hk,
  perms = ion_crossobject_addcov.operm, alpha = c(0.05), 
  format= "tabByChr"
  ) 

# transform the summary to tibble
scanone_summary_add_cov <- summary.to.tibble(
  CROSS = ion_crossobject, 
  SCANONE = ion_crossobject_addcov.out.hk,
  PERMS = ion_crossobject_addcov.operm, 
  alpha = 0.1
  )

# Export the summary as a csv
write_csv(
  scanone_summary_add_cov,
  "output/tables/scanone_summary/scanone_summary_add_cov.csv"
)

# Create dataframe for lodplots
ion_add_cov_lodplot_df <- lodplot_map_df(
  SCANONE = ion_crossobject_addcov.out.hk,
  SUMMARY = scanone_summary_add_cov,
  PERMS = ion_crossobject_addcov.operm,
  alpha = 0.1,
  ci.funciton = "lodint"
  )

# lodplots for all the phenotypes
ion_add_cov_lodplots_all <- ion_add_cov_lodplot_df %>%
  group_by(phenotype) %>%
  nest %>%
  mutate(
    title = paste0(
      "QTL analysis for additive covariate for ",
      phenotype)) %>%
  mutate(
    plot = map2(
      .x = data, 
      .y = title,
      .f = ~ .x %>% 
        {
        ggplot(
          data =., 
          aes(x = cM, 
              y = lod, 
              group = interaction(chr, alpha)
          )
          ) +
            geom_line(size = 0.75, alpha = 0.3) +
          geom_line(
            aes(alpha = alpha),
            size = 0.75
            ) +
          geom_ribbon(
            aes(ymin = 0, ymax = y), 
            fill = "red", 
            alpha = 0.8
            ) +
          geom_hline( 
            aes(yintercept = unique(threshold)),
            color = "red",
            linetype = 2
            ) +
          scale_alpha_identity() +
          scale_x_continuous(
            name = "chr", 
            breaks = unique(.$breaks), 
            labels = unique(as.character((.$chr)))
            ) +
            theme(
              panel.background = 
                element_rect(
                fill = "white",
                colour = "grey50")) +
            ggtitle(.y)
      }
    )
  ) 

# # create directory for plots
# dir.create(file.path(paste0(getwd(), "/output/plots"),
#                      "scanone_lodplots"))
p <- ion_add_cov_lodplots_all$plot

# Saving plots into a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_add_cov_lodplots_all.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# Only significant plots
ion_add_cov_lodplots_signif <- ion_add_cov_lodplots_all %>%
  semi_join(scanone_summary_add_cov)

p <- ion_add_cov_lodplots_signif$plot

# Saving plots into a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_add_cov_lodplots_signif.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# all significant plots in the same page
ion_add_cov_lodplots_signif_single <- ion_add_cov_lodplots_signif %>%
  select(phenotype, data) %>%
  unnest(data) %>% {
    ggplot(data =., 
           aes(x = cM, y = lod, group = interaction(chr, alpha))) +
      geom_line(size = 0.75, alpha = 0.3) +
      geom_line(aes(alpha = alpha), size = 0.75) +
      geom_ribbon(aes(ymin = 0, ymax = y), fill = "red", alpha = 0.8 ) +
      geom_hline(aes(yintercept = threshold), color = "red",
                 linetype = 2) +
      scale_alpha_identity() +
      scale_x_continuous(name = "chr", 
                         breaks = unique(.$breaks),
                         labels = unique(as.character((.$chr)))) +
  theme(panel.background = element_rect( fill = "white",
                                         colour = "grey50")) +
  ggtitle("QTL analysis sig peaks additive covariate") +
  facet_grid(phenotype ~ .)
      }

ggsave(
   filename =
     "output/plots/scanone_lodplots/ion_add_cov_lodplots_signif_single.pdf", 
   plot = ion_add_cov_lodplots_signif_single, 
   width = 15, 
   height = 18
)
```

QTL analysis with an interactive covariate model

```{r}
##################################################
# Using scanone aplying an interactive covariate #
##################################################

ion_crossobject_intcov.out.hk <- 
  scanone(ion_crossobject, 
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          addcovar = hun_cov,
          intcovar = hun_cov
          )

# save scanone
save(
  list = c("ion_crossobject_intcov.out.hk"), 
  file = "output/qtl_objects/scanone/ion_crossobject_intcov.out.hk.RData"
)

load("output/qtl_objects/scanone/ion_crossobject_intcov.out.hk.RData")
# Permutation test
ion_crossobject_intcov.operm <- 
  scanone(ion_crossobject, 
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          addcovar = hun_cov,
          intcovar = hun_cov,
          n.perm = 1000
          )

# save permutation test
save(
  list = c("ion_crossobject_intcov.operm"), 
  file = "output/qtl_objects/permutations/ion_crossobject_intcov.operm.RData"
)

load("output/qtl_objects/permutations/ion_crossobject_intcov.operm.RData")

# Summary 
summary(
  ion_crossobject_intcov.out.hk, 
  perms = ion_crossobject_intcov.operm, 
  alpha = c(0.1), 
  format= "tabByChr"
  )

# transform the summary to tibble
scanone_summary_int_cov <- summary.to.tibble(
  CROSS = ion_crossobject, 
  SCANONE = ion_crossobject_intcov.out.hk,
  PERMS = ion_crossobject_intcov.operm, 
  alpha = 0.1
  )

# Export the summary as a csv
write_csv(
  scanone_summary_int_cov,
  "output/tables/scanone_summary/scanone_summary_int_cov.csv"
)

# Create dataframe for lodplots
ion_int_cov_lodplot_df <- lodplot_map_df(
  SCANONE = ion_crossobject_intcov.out.hk,
  SUMMARY = scanone_summary_int_cov,
  PERMS = ion_crossobject_intcov.operm,
  alpha = 0.1,
  ci.funciton = "lodint"
  )

# lodplots for all the phenotypes
ion_int_cov_lodplots_all <- ion_int_cov_lodplot_df %>%
  group_by(phenotype) %>%
  nest %>%
  mutate(
    title = paste0(
      "QTL analysis for intitive covariate for ",
      phenotype)) %>%
  mutate(
    plot = map2(
      .x = data, 
      .y = title,
      .f = ~ .x %>% 
        {
        ggplot(
          data =., 
          aes(x = cM, 
              y = lod, 
              group = interaction(chr, alpha)
          )
          ) +
            geom_line(size = 0.75, alpha = 0.3) +
          geom_line(
            aes(alpha = alpha),
            size = 0.75
            ) +
          geom_ribbon(
            aes(ymin = 0, ymax = y), 
            fill = "red", 
            alpha = 0.8
            ) +
          geom_hline( 
            aes(yintercept = unique(threshold)),
            color = "red",
            linetype = 2
            ) +
          scale_alpha_identity() +
          scale_x_continuous(
            name = "chr", 
            breaks = unique(.$breaks), 
            labels = unique(as.character((.$chr)))
            ) +
            theme(
              panel.background = 
                element_rect(
                fill = "white",
                colour = "grey50")) +
            ggtitle(.y)
      }
    )
  ) 

# create directory for plots
dir.create(file.path(paste0(getwd(), "/output/plots"),
                     "scanone_lodplots"))
p <- ion_int_cov_lodplots_all$plot

# Saving plots into a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_int_cov_lodplots_all.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# Only significant plots
ion_int_cov_lodplots_signif <- ion_int_cov_lodplots_all %>%
  semi_join(scanone_summary_int_cov)

p <- ion_int_cov_lodplots_signif$plot

# Saving plots into a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_int_cov_lodplots_signif.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# all significant plots in the same page
ion_int_cov_lodplots_signif_single <- ion_int_cov_lodplots_signif %>%
  select(phenotype, data) %>%
  unnest(data) %>% {
    ggplot(data =., 
           aes(x = cM, y = lod, group = interaction(chr, alpha))) +
      geom_line(size = 0.75, alpha = 0.3) +
      geom_line(aes(alpha = alpha), size = 0.75) +
      geom_ribbon(aes(ymin = 0, ymax = y), fill = "red", alpha = 0.8 ) +
      geom_hline(aes(yintercept = threshold), color = "red",
                 linetype = 2) +
      scale_alpha_identity() +
      scale_x_continuous(name = "chr", 
                         breaks = unique(.$breaks),
                         labels = unique(as.character((.$chr)))) +
  theme(panel.background = element_rect( fill = "white",
                                         colour = "grey50")) +
  ggtitle("QTL analysis sig peaks intitive covariate") +
  facet_grid(phenotype ~ .)
      }

ggsave(
   filename =
     "output/plots/scanone_lodplots/ion_int_cov_lodplots_signif_single.pdf", 
   plot = ion_int_cov_lodplots_signif_single, 
   width = 15, 
   height = 18
)
```

Additive and Interactive models comparison

```{r}
# Comparing models and looking for evidence for interaction between QTLs and HUN covariate

# scanone
ion_int_add.out.hk <- c(
  ion_crossobject_intcov.out.hk,
  ion_crossobject_intcov.out.hk - ion_crossobject_addcov.out.hk,
  labels = c("f", "i")
)

# save scanone
save(
  list = c("ion_int_add.out.hk"), 
  file = "output/qtl_objects/scanone/ion_int_add.out.hk.RData"
)

# permutations
ion_int_add.operm <- cbind(
  ion_crossobject_intcov.operm,
  ion_crossobject_intcov.operm - ion_crossobject_addcov.operm,
  labels = c("f", "i")
)

# save permutation test
save(
  list = c("ion_int_add.operm"), 
  file = "output/qtl_objects/permutations/ion_int_add.operm.RData"
)

load("output/qtl_objects/permutations/ion_int_add.operm.RData")

# Summary
summary(
  ion_int_add.out.hk, 
  perms = ion_int_add.operm, 
  alpha = c(0.05), 
  format = "tabByChr"
  )

# transform the summary to tibble
scanone_summary_int_add_cov <- summary.to.tibble(
  CROSS = ion_crossobject, 
  SCANONE = ion_int_add.out.hk,
  PERMS = ion_int_add.operm, 
  alpha = 0.1
  )

# Export the summary as a csv
write_csv(
  scanone_summary_int_add_cov,
  "output/tables/scanone_summary/scanone_summary_int_add_cov.csv"
)

ions_interesting_qtl_addint <- scanone_summary_int_add_cov %>%
  arrange(phenotype) %>%
  separate(phenotype, into = c("phenotype2", "model"), sep = "\\.",
           remove = F) %>%
  group_by(phenotype2, chr) %>%
  filter(dplyr::n() != 1)

# Create dataframe for lodplots
ion_int_add_cov_lodplot_df <- lodplot_map_df(
  SCANONE = ion_int_add.out.hk,
  SUMMARY = scanone_summary_int_add_cov,
  PERMS = ion_int_add.operm,
  alpha = 0.1,
  ci.funciton = "lodint"
  )

# lodplots for all the phenotypes
ion_int_add_cov_lodplots_all <- ion_int_add_cov_lodplot_df %>%
  group_by(phenotype) %>%
  nest %>%
  arrange(phenotype) %>%
  mutate(
    title = paste0(
      "QTL analysis for int_additive covariate for ",
      phenotype)) %>%
  mutate(
    plot = map2(
      .x = data, 
      .y = title,
      .f = ~ .x %>% 
        {
        ggplot(
          data =., 
          aes(x = cM, 
              y = lod, 
              group = interaction(chr, alpha)
          )
          ) +
            geom_line(size = 0.75, alpha = 0.3) +
          geom_line(
            aes(alpha = alpha),
            size = 0.75
            ) +
          geom_ribbon(
            aes(ymin = 0, ymax = y), 
            fill = "red", 
            alpha = 0.8
            ) +
          geom_hline( 
            aes(yintercept = unique(threshold)),
            color = "red",
            linetype = 2
            ) +
          scale_alpha_identity() +
          scale_x_continuous(
            name = "chr", 
            breaks = unique(.$breaks), 
            labels = unique(as.character((.$chr)))
            ) +
            theme(
              panel.background = 
                element_rect(
                fill = "white",
                colour = "grey50")) +
            ggtitle(.y)
      }
    )
  ) 


p <- ion_int_add_cov_lodplots_all$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_int_add_cov_lodplots_all.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# Only significant plots
ion_int_add_cov_lodplots_signif <- ion_int_add_cov_lodplots_all %>%
  semi_join(scanone_summary_int_add_cov)

p <- ion_int_add_cov_lodplots_signif$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_int_add_cov_lodplots_signif.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)


# all significant plots in the same page

ion_int_add_cov_lodplots_interesting <- ion_int_add_cov_lodplots_all %>%
  semi_join(ions_interesting_qtl_addint) %>%
  select(phenotype, data) %>%
  unnest(data) %>% {
    ggplot(data =., 
           aes(x = cM, y = lod, group = interaction(chr, alpha))) +
      geom_line(size = 0.75, alpha = 0.3) +
      geom_line(aes(alpha = alpha), size = 0.75) +
      geom_ribbon(aes(ymin = 0, ymax = y), fill = "red", alpha = 0.8 ) +
      geom_hline(aes(yintercept = threshold), color = "red",
                 linetype = 2) +
      scale_alpha_identity() +
      scale_x_continuous(name = "chr", 
                         breaks = unique(.$breaks),
                         labels = unique(as.character((.$chr)))) +
  theme(panel.background = element_rect( fill = "white",
                                         colour = "grey50")) +
  ggtitle("QTL analysis sig peaks int_additive covariate") +
  facet_grid(phenotype ~ .)
      }

ggsave(
   filename =
     "output/plots/scanone_lodplots/ion_int_add_cov_lodplots_interesting.pdf", 
   plot = ion_int_add_cov_lodplots_interesting, 
   width = 15, 
   height = 18
)

```

Separate analysis on WT families

```{r}
################################
# Analysis on WT families ONLY #
################################

ion_crossobject_WT <- subset(ion_crossobject, ind  = hun_cov == 0)

# scanone for wt individuals
ion_crossobject_WT.out.hk <- 
  scanone(ion_crossobject_WT,
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col
          )

# save scanone
save(
  list = c("ion_crossobject_WT.out.hk"), 
  file = "output/qtl_objects/scanone/ion_crossobject_WT.out.hk.RData"
)

# Permutation test (1000 permutations) subsetting for wt individuals
ion_crossobject_WT.operm <- 
  scanone(subset(ion_crossobject, ind  = hun_cov == 0),
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          n.perm = 1000
          )
# save permutation test
save(
  list = c("ion_crossobject_WT.operm"), 
  file = "output/qtl_objects/permutations/ion_crossobject_WT.operm.RData"
)

load("output/qtl_objects/permutations/ion_crossobject_WT.operm.RData")
# Summary, alpha = 0.05
summary(
  ion_crossobject_WT.out.hk,
  perms = ion_crossobject_WT.operm,
  alpha = c(0.1), 
  format= "tabByChr"
  ) 

# transform the summary to tibble
scanone_summary_WT <- summary.to.tibble(
  CROSS = ion_crossobject_WT, 
  SCANONE = ion_crossobject_WT.out.hk,
  PERMS = ion_crossobject_WT.operm, 
  alpha = 0.1
  )

# Export the summary as a csv
write_csv(
  scanone_summary_int_add_cov,
  "output/tables/scanone_summary/scanone_summary_WT.csv"
)

# Create dataframe for lodplots
ion_WT_lodplot_df <- lodplot_map_df(
  SCANONE = ion_crossobject_WT.out.hk,
  SUMMARY = scanone_summary_WT,
  PERMS = ion_crossobject_WT.operm,
  alpha = 0.1,
  )

# lodplots for all the phenotypes
ion_WT_lodplots_all <- ion_WT_lodplot_df %>%
  group_by(phenotype) %>%
  nest %>%
  arrange(phenotype) %>%
  mutate(
    title = paste0(
      "QTL analysis for int_additive covariate for ",
      phenotype)) %>%
  mutate(
    plot = map2(
      .x = data, 
      .y = title,
      .f = ~ .x %>% 
        {
        ggplot(
          data =., 
          aes(x = cM, 
              y = lod, 
              group = interaction(chr, alpha)
          )
          ) +
            geom_line(size = 0.75, alpha = 0.3) +
          geom_line(
            aes(alpha = alpha),
            size = 0.75
            ) +
          geom_ribbon(
            aes(ymin = 0, ymax = y), 
            fill = "red", 
            alpha = 0.8
            ) +
          geom_hline( 
            aes(yintercept = unique(threshold)),
            color = "red",
            linetype = 2
            ) +
          scale_alpha_identity() +
          scale_x_continuous(
            name = "chr", 
            breaks = unique(.$breaks), 
            labels = unique(as.character((.$chr)))
            ) +
            theme(
              panel.background = 
                element_rect(
                fill = "white",
                colour = "grey50")) +
            ggtitle(.y)
      }
    )
  ) 


p <- ion_WT_lodplots_all$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_WT_lodplots_all.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# Only significant plots
ion_WT_lodplots_signif <- ion_WT_lodplots_all %>%
  semi_join(scanone_summary_WT)

p <- ion_WT_lodplots_signif$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_WT_lodplots_signif.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# all significant plots in the same page

ion_WT_lodplots_signif_single <- ion_WT_lodplots_all %>%
  semi_join(scanone_summary_WT) %>%
  select(phenotype, data) %>%
  unnest(data) %>% {
    ggplot(data =., 
           aes(x = cM, y = lod, group = interaction(chr, alpha))) +
      geom_line(size = 0.75, alpha = 0.3) +
      geom_line(aes(alpha = alpha), size = 0.75) +
      geom_ribbon(aes(ymin = 0, ymax = y), fill = "red", alpha = 0.8 ) +
      geom_hline(aes(yintercept = threshold), color = "red",
                 linetype = 2) +
      scale_alpha_identity() +
      scale_x_continuous(name = "chr", 
                         breaks = unique(.$breaks),
                         labels = unique(as.character((.$chr)))) +
  theme(panel.background = element_rect( fill = "white",
                                         colour = "grey50")) +
  ggtitle("QTL analysis sig peaks int_additive covariate") +
  facet_grid(phenotype ~ .)
      }

ggsave(
   filename =
     "output/plots/scanone_lodplots/ion_WT_lodplots_signif_single.pdf", 
   plot = ion_WT_lodplots_signif_single, 
   width = 15, 
   height = 18
)

```

Separate analysis on HUN families

```{r}

#################################
# Analysis on HUN families ONLY #
#################################
ion_crossobject_HUN <- subset(ion_crossobject, ind  = hun_cov == 1)

# scanone for HUN individuals
ion_crossobject_HUN.out.hk <- 
  scanone(ion_crossobject_HUN,
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col
          )

# save scanone
save(
  list = c("ion_crossobject_HUN.out.hk"), 
  file = "output/qtl_objects/scanone/ion_crossobject_HUN.out.hk.RData"
)

# Permutation test (1000 permutations) subsetting for HUN individuals
ion_crossobject_HUN.operm <- 
  scanone(subset(ion_crossobject, ind  = hun_cov == 1),
          method = "hk",
          pheno.col = ion_pheno.col$pheno.col, 
          n.perm = 1000
          )
# save permutation test
save(
  list = c("ion_crossobject_HUN.operm"), 
  file = "output/qtl_objects/permutations/ion_crossobject_HUN.operm.RData"
)

load(file = "output/qtl_objects/permutations/ion_crossobject_HUN.operm.RData")

# Summary, alpha = 0.05
summary(
  ion_crossobject_HUN.out.hk,
  perms = ion_crossobject_HUN.operm,
  alpha = c(0.1), 
  format= "tabByChr"
  ) 

# transform the summary to tibble
scanone_summary_HUN <- summary.to.tibble(
  CROSS = ion_crossobject_HUN, 
  SCANONE = ion_crossobject_HUN.out.hk,
  PERMS = ion_crossobject_HUN.operm, 
  alpha = 0.1
  )

# Export the summary as a csv
write_csv(
  scanone_summary_int_add_cov,
  "output/tables/scanone_summary/scanone_summary_HUN.csv"
)

# Create dataframe for lodplots
ion_HUN_lodplot_df <- lodplot_map_df(
  SCANONE = ion_crossobject_HUN.out.hk,
  SUMMARY = scanone_summary_HUN,
  PERMS = ion_crossobject_HUN.operm,
  alpha = 0.1,
  )

# lodplots for all the phenotypes
ion_HUN_lodplots_all <- ion_HUN_lodplot_df %>%
  group_by(phenotype) %>%
  nest %>%
  arrange(phenotype) %>%
  mutate(
    title = paste0(
      "QTL analysis for int_additive covariate for ",
      phenotype)) %>%
  mutate(
    plot = map2(
      .x = data, 
      .y = title,
      .f = ~ .x %>% 
        {
        ggplot(
          data =., 
          aes(x = cM, 
              y = lod, 
              group = interaction(chr, alpha)
          )
          ) +
            geom_line(size = 0.75, alpha = 0.3) +
          geom_line(
            aes(alpha = alpha),
            size = 0.75
            ) +
          geom_ribbon(
            aes(ymin = 0, ymax = y), 
            fill = "red", 
            alpha = 0.8
            ) +
          geom_hline( 
            aes(yintercept = unique(threshold)),
            color = "red",
            linetype = 2
            ) +
          scale_alpha_identity() +
          scale_x_continuous(
            name = "chr", 
            breaks = unique(.$breaks), 
            labels = unique(as.character((.$chr)))
            ) +
            theme(
              panel.background = 
                element_rect(
                fill = "white",
                colour = "grey50")) +
            ggtitle(.y)
      }
    )
  ) 


p <- ion_HUN_lodplots_all$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_HUN_lodplots_all.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# Only significant plots
ion_HUN_lodplots_signif <- ion_HUN_lodplots_all %>%
  semi_join(scanone_summary_HUN)

p <- ion_HUN_lodplots_signif$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_HUN_lodplots_signif.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)

# all significant plots in the same page

ion_HUN_lodplots_signif <- ion_HUN_lodplots_all %>%
  semi_join(ion_HUN_lodplots_signif) %>%
  select(phenotype, data) %>%
  unnest(data) %>% {
    ggplot(data =., 
           aes(x = cM, y = lod, group = interaction(chr, alpha))) +
      geom_line(size = 0.75, alpha = 0.3) +
      geom_line(aes(alpha = alpha), size = 0.75) +
      geom_ribbon(aes(ymin = 0, ymax = y), fill = "red", alpha = 0.8 ) +
      geom_hline(aes(yintercept = threshold), color = "red",
                 linetype = 2) +
      scale_alpha_identity() +
      scale_x_continuous(name = "chr", 
                         breaks = unique(.$breaks),
                         labels = unique(as.character((.$chr)))) +
  theme(panel.background = element_rect( fill = "white",
                                         colour = "grey50")) +
  ggtitle("QTL analysis sig peaks int_additive covariate") +
  facet_grid(phenotype ~ .)
      }

ggsave(
   filename =
     "output/plots/scanone_lodplots/ion_HUN_lodplots_signif.pdf", 
   plot = ion_HUN_lodplots_signif, 
   width = 15, 
   height = 18
)

```

QTLs detected in all the separated analysis

```{r}
# combine all scanone summaries
qtl_detected_all <- bind_rows(
  scanone_summary_add_cov %>% mutate(model = "add"),
  scanone_summary_int_cov %>% mutate(model = "int"),
  ions_interesting_qtl_addint %>% mutate(model = "addint"),
  scanone_summary_WT %>% mutate(model = "WT"),
  scanone_summary_HUN %>% mutate(model = "HUN"),
  MQM_ion_hun_summary  %>% mutate(model = "MQM")
  ) %>%
  separate(phenotype, into = c("phenotype", "sufix"), sep = "\\.") %>%
  mutate(model = ifelse(!is.na(sufix), 
                        paste0(model, ".", sufix), 
                        model)) %>%
  select(-c(sufix, phenotype2)) %>%
  select(all_of(c(names(scanone_summary_add_cov), "model")))

# Combine all information for plots
qtl_detected_all_lodplot <- 
  bind_rows(
  ion_add_cov_lodplots_all %>% mutate(model = "add"),
  ion_int_cov_lodplots_all %>% mutate(model = "int"),
  ion_int_add_cov_lodplots_all %>% mutate(model = "addint"),
  ion_WT_lodplots_all %>% mutate(model = "WT"),
  ion_HUN_lodplots_all %>% mutate(model = "HUN")
  ) %>%
  select(phenotype, data, model) %>%
  separate(phenotype, into = c("phenotype", "sufix"), sep = "\\.") %>%
  mutate(model = ifelse(!is.na(sufix),
                        paste0(model, ".", sufix),
                        model)) %>%
  select(-sufix) %>%
  semi_join(qtl_detected_all) %>% distinct

# stablish a color scheme for the models in the plots

model_color_scheme <- 
  c(
    "add" = "#E41A1C", "addint.f" = "#377EB8", "addint.i" = "#4DAF4A",
    "HUN" = "#984EA3", "int" = "#FF7F00", "WT" = "#FFFF33"
    )

ion_lodplots_all_df <- qtl_detected_all_lodplot %>%
  separate(phenotype, into = c("ion", "tissue"), sep = "_") %>%
  mutate(
    tissue = factor(
      tissue,
      levels = c("leaf", "seed", "ratio", "mean")
      )) %>%
  unnest(data) %>%
  group_by(ion) %>%
  nest %>%
  ungroup 

function_plots_a <- function(data, ion) {
  
  plot <- data %>%
    {
      ggplot(
        data =.,
        aes( x = cM, 
             y = lod, 
             group = interaction(chr, model),
             color = model)) +
        geom_line(size = 1,
                  alpha =  0.3) +
        geom_line(aes(alpha = alpha),
                  size = 1) +
        scale_alpha_identity() +
        scale_color_manual(values = model_color_scheme) +
        scale_x_continuous(name = NULL, 
                           breaks = unique(.$breaks), 
                           labels = unique(as.character((.$chr)))) +
        theme(panel.background = element_rect( fill = "white",
                                               colour = "grey50"),
              legend.position = "top",
              plot.title = element_text(hjust = 0.5)) +
        ylab("LOD") +
        facet_grid(tissue ~ .) +
        ggtitle(paste0("Ion: ", ion))
    }
  return(plot)
}


ion_lodplots_all_plot <- ion_lodplots_all_df %>% 
  mutate(plot = map2(.x = data, 
                     .y = ion, 
                     .f = ~ function_plots_a(data = .x, ion = .y)))

p <- ion_lodplots_all_plot$plot

# Saving plots int_addo a single pdf
ggsave(
   filename = 
     "output/plots/scanone_lodplots/ion_lodplots_all_plot.pdf", 
   plot = gridExtra::marrangeGrob(p, nrow=1, ncol=1), 
   width = 15, height = 9
)
```

map of QTL locations across chromosomes

```{r}

model_line_scheme <- c("HUN" = "dotdash", "add" = "F1", "int" = "dotted",
                       "WT" = "longdash", "addint.i" = "solid",
                       "addint.f" = "twodash", "MQM" = "12345678")

phenotypes_color_scheme_df <- ion_pheno.col %>% 
  separate(phenotype, into = c("ion", "tissue")) %>%
  mutate(tissue = factor(
    tissue, 
    levels = c("leaf", "seed", "ratio", "mean"
               )
    )) %>%
  arrange(ion, tissue) %>% 
  mutate(phenotype = paste0(ion, "_", tissue)) %>%
  mutate(color = parula(80)) %>%
  mutate(phenotype = as_factor(phenotype))

phenotypes_color_scheme <- phenotypes_color_scheme_df$color

names(phenotypes_color_scheme) <- levels(
  phenotypes_color_scheme_df$phenotype
  )

phen_cs <- scale_colour_manual(
  values = phenotypes_color_scheme,
  drop = T
  )

chr_plot_base_df <- qtl_detected_all_lodplot %>%
  unnest(data) %>%
  select(chr, pos) %>%
  group_by(chr) %>%
  summarise(min = min(pos),
            max = max(pos)) 

chr_plot_phenotypes_df <- qtl_detected_all %>%
  arrange(chr, phenotype) %>%
  select(chr, phenotype, pos, ci.low, ci.high, model) %>%
  mutate(chr = as.integer(chr)) %>%
  arrange(chr) %>%
  group_by(chr) %>%
  mutate(id = row_number()) %>%
  mutate(comp = ifelse(id %% 2 != 0, 1, -1)) %>%
  arrange(chr, desc(comp), phenotype) %>%
  group_by(chr, comp) %>%
  mutate(id2 = row_number()) %>%
  mutate(x = 1 + 0.05 * comp * id2) %>%
  group_by(chr) %>%
  filter(x != chr) %>%
  select(-c(id:id2)) %>%
  ungroup %>%
  pivot_longer(contains("ci")) %>%
  mutate(
    phenotype = factor(
      phenotype, 
      levels = levels(phenotypes_color_scheme_df$phenotype)
      ))

ion_chrom_plot <- chr_plot_phenotypes_df %>%
  left_join(chr_plot_base_df) %>%
  mutate(chr = factor(chr, levels = c(1:10))) %>%
  pivot_longer(
    c(min, max),
      names_to = "dist", values_to ="dist_value") %>%
  ggplot() +
    geom_line(
      aes(
        x = chr, 
        y = dist_value), 
      size = 1.25) +
    geom_line(
      aes(
        x = x, 
        y = value, 
        group = interaction(chr, phenotype, model),
        color = phenotype,
        linetype = model),
      size = 1
         ) +
    scale_y_continuous(trans = "reverse") +
    scale_linetype_manual(values = model_line_scheme) +
    scale_color_manual(
      values = phenotypes_color_scheme,
      drop = T, 
      breaks = chr_plot_phenotypes_df$phenotype %>% 
        unique %>% 
        sort
      ) +
    theme(
      panel.background = 
          element_rect(
            fill = "white", 
            colour ="grey50"),
      plot.title = element_text(hjust = 0.5),
      legend.position = "top"
      ) +
  xlab(NULL) +
  facet_grid(. ~ chr, scales = "free_x") +
  ggtitle("QTLs detected with scanone in the analysis") + 
  guides(col = guide_legend(ncol = 10))

# dir.create(file.path(paste0(getwd(), "/output/plots"), "chrom_plot"))
ggsave(
   filename ="output/plots/chrom_plot/chrom_plot_all.pdf", 
   plot = ion_chrom_plot,
   width = 15, height = 9
)
```

Effect plots for QTLs identified in HUN and WT models

```{r}
genotypes_scheme_color <- c("CML312" = "#e77e39", 
                            "HET" = "#9a3a28",
                            "W22" = "#2e1342")

ion_pheno.col2 <- ion_pheno.col %>%
  separate(phenotype, into = c("ion", "tissue"), remove = F)

ion_effectplot_df_pre <- qtl_detected_all %>%
  select(phenotype, chr, pos, model) %>%
  separate(phenotype, into = c("ion"), sep = "_", 
           remove = F) %>%
  distinct %>%
  left_join(ion_pheno.col2 %>% select(-phenotype)) %>%
  mutate(tissue = factor(
    tissue,
    levels = c("leaf", "seed", "ratio", "mean"))) %>%
  separate(phenotype, into = c("phenotype", "sufix"), sep = "\\.") %>%
  mutate(model = ifelse(!is.na(sufix),
                        paste0(model, ".", sufix), 
                        model)) %>%
  select(-c(sufix)) %>%
  arrange(chr, pos, phenotype)  %>%
  left_join(ion_pheno.col) %>%
  mutate(pos = round(pos, 1),
         qtl = paste0(chr, "@", pos)) %>%
  mutate(crossobject = case_when(
    model == "HUN" ~ list(ion_crossobject_HUN),
    model == "WT" ~ list(ion_crossobject_WT),
    T ~ list(ion_crossobject))) 


# Effect plots forn hun, wt and MQM (without hun cov interaction)
ion_effectplot_df_HUN_WT <- ion_effectplot_df_pre %>%
  filter(model %in% c("WT", "HUN", "MQM")) %>%
  mutate(ep = pmap(
    .l = .,
    .f = ~ with( 
      list(...),
      effectplot(cross = crossobject, pheno.col = pheno.col, 
                 mname1 = qtl, draw = F))))
  
ion_effectplot_df_HUN_WT_ep <- ion_effectplot_df_HUN_WT %>%
  select(model, phenotype, ion, chr, tissue, qtl, model, ep) %>%
  mutate(pheno2 = paste0(ion, "_", tissue)) %>%
  mutate(is.QTL = ifelse(phenotype == pheno2, TRUE, FALSE),
         is.QTL = as.factor(is.QTL)) %>%
  select(-c(pheno2)) %>%
  mutate(ep_df = map( 
    .x = ep,
    .f = ~ map_df(.x =., .f = ~enframe(.x), .id = "what") %>%
      mutate(name = str_extract(name, "[A-Z0-9]{1,}$")) %>%
      pivot_wider(names_from = what, values_from = value) %>%
      mutate(name = case_when(name == "CMLCML" ~ "CML312",
                              name == "CMLW22" ~ "HET",
                              name == "W22W22" ~ "W22",
                              T ~ NA_character_)))) %>%
  select(-ep) %>%
  unnest(ep_df) 

# Remove this object that's insanely heavy
rm(ion_effectplot_df_HUN_WT)

# effectplots 
ion_effectplot_df_HUN_WT_ep_plots <- ion_effectplot_df_HUN_WT_ep %>%
  mutate(title = paste("Effect plots for ", 
                       phenotype, 
                       "QTL in position", 
                       qtl)) %>%
  group_by(model, phenotype, chr, title) %>%
  nest %>%
  separate(phenotype, into = c("ion", NA), remove = F) %>%
  mutate(plot = map2(
    .x = data,
    .y = title,
    .f = ~ .x %>%
      ggplot(data =., aes(x = name, y = Means, color = name))  +
      geom_errorbar(aes(ymin = Means - SEs , ymax = Means + SEs),
                    width = 0.2, 
                    size = 1) +
      geom_point(size = 2, shape = 22, fill = "white") +
      scale_color_manual(name = "Genotype",
                         values = genotypes_scheme_color) +
      xlab(NULL) +
      ylab("Concentration (ppm)") +
      theme(
        panel.background = element_rect(fill = "white", 
                                        colour ="grey50"),
        plot.title = element_text(hjust = 0.5),
        text = element_text(size=20)) +
      facet_wrap(. ~ tissue + is.QTL, 
                 nrow = 1,
                 scales = "free_y",
                 labeller = ggplot2::labeller(tissue = label_value,
                                              is.QTL = label_both)) +
      ggtitle(.y))) %>%
  group_by(model, ion, chr) %>%
  summarise(plot = list(plot)) %>%
  mutate(title = paste0("Ion: ", ion)) %>%
  mutate(wrap = map2(
    .x = plot, 
    .y = title,
    .f = ~ patchwork::wrap_plots(.x, guides = "collect", ncol = 1) +
      patchwork::plot_annotation( 
        title = .y,
        theme = theme(plot.title = element_text(hjust = 0.5))))) %>%
  ungroup

p1 <- ion_effectplot_df_HUN_WT_ep_plots %>%
  filter(model == "HUN") %>%
  .$wrap

p2 <- ion_effectplot_df_HUN_WT_ep_plots %>%
  filter(model == "WT") %>%
  .$wrap

p3 <- ion_effectplot_df_HUN_WT_ep_plots %>%
  filter(model == "MQM") %>%
  .$wrap


# # # create directory for plots
# dir.create(file.path(paste0(getwd(), "/output/plots"), "effectplots"))

# Saving plots into a single pdf
pdf("output/plots/effectplots/HUN_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p1, ~ print(.x))
dev.off()

# Saving plots into a single pdf
pdf("output/plots/effectplots/WT_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p2, ~ print(.x))
dev.off()

# Saving plots into a single pdf
pdf("output/plots/effectplots/MQM_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p3, ~ print(.x))
dev.off()


```

Effect plots for QTLs identified in add and int models

```{r}

# Effect plots forn add, int and adint models (hun cov interaction)
ion_effectplot_df_cov <- ion_effectplot_df_pre %>%
  filter(! model %in% c("WT", "HUN", "MQM")) %>%
  mutate(ep = pmap(
    .l = .,
    .f = ~ with( 
      list(...),
      effectplot(cross = crossobject, pheno.col = pheno.col, 
                 mname1 = "5_2269274", mname2 = qtl, draw = F))))
  
ion_effectplot_df_cov2 <- ion_effectplot_df_cov %>%
  select(-crossobject) %>%
  mutate(
    epdf = map(
      .x = ep,
      .f = ~ map_df(
        .x =., 
        .f = ~ as_tibble(.x, rownames = "hun"), 
        .id = "what") %>%
  pivot_longer(-c(what, hun)) %>%
  pivot_wider(names_from = what, values_from = value) %>%
  filter(!grepl("CMLW22", hun)) %>%
  mutate(across(c(hun, name), ~ str_extract(., "[A-Z0-9]{1,}$"))) %>%
  mutate(hun = ifelse(hun == "CMLCML", "WT", "HUN"),
         name = case_when(name == "CMLCML" ~ "CML312",
                          name == "CMLW22" ~ "HET",
                          name == "W22W22" ~ "W22",
                          T ~ NA_character_)) %>%
  mutate(hun = factor(hun, levels = c("WT", "HUN"))))) 

cov_effecplots_function <- function(data, title) {
  
  PLOT <- data %>%
    ggplot(
    data =., 
    aes(x = hun, 
        y = Means,
        group =  name,
        color = name
        )
    ) +
  geom_errorbar(
    aes(
      ymin = Means - SEs,
      ymax = Means + SEs
      ),
    width = 0.2, 
    size = 1
    ) +
  geom_line(
    size = 1
    ) +
  geom_point(
    size = 2, 
    shape = 22, 
    fill = "white"
    ) +
  scale_color_manual(
    name = "Genotype",
    values = genotypes_scheme_color
    ) +
  xlab(NULL) +
  ylab("Concentration (ppm)") +
  theme(
        panel.background = element_rect(
          fill = "white", 
          colour ="grey50"
          ),
        plot.title = element_text(hjust = 0.5),
        text = element_text(size=20)
        ) +
      facet_wrap(
        . ~ tissue + is.QTL,
        nrow = 1,
        scales = "free_y",
        labeller = ggplot2::labeller(tissue = label_value,
                                     is.QTL = label_both
                                     )
        ) +
    ggtitle(title)
  
  return(PLOT)

  
}

ion_effectplot_df_cov_ep_plots <- ion_effectplot_df_cov2 %>%
  select(model, phenotype, ion, chr, tissue, qtl, model, epdf) %>%
  mutate(pheno2 = paste0(ion, "_", tissue)) %>%
  mutate(is.QTL = ifelse(phenotype == pheno2, TRUE, FALSE),
         is.QTL = as.factor(is.QTL)) %>%
  select(-c(pheno2)) %>%
  unnest(epdf) %>%
  mutate(title = paste("Effect plots for ", 
                       phenotype, 
                       "QTL in position", 
                       qtl)) %>%
  group_by(model, phenotype, chr, title) %>%
  nest %>%
  separate(phenotype, into = c("ion", NA), remove = F) %>%
  mutate(plot = map2(
    .x = data,
    .y = title,
    .f = ~ cov_effecplots_function(.x, .y)
    )) %>%
  group_by(model, ion, chr) %>%
  summarise(plot = list(plot)) %>%
  mutate(title = paste0("Ion: ", ion)) %>%
  mutate(wrap = map2(
    .x = plot, 
    .y = title,
    .f = ~ patchwork::wrap_plots(.x, guides = "collect", ncol = 1) +
      patchwork::plot_annotation( 
        title = .y,
        theme = theme(plot.title = element_text(hjust = 0.5))))) %>%
  ungroup 
  

p1 <- ion_effectplot_df_cov_ep_plots %>%
  filter(model == "add") %>%
  .$wrap

p2 <- ion_effectplot_df_cov_ep_plots %>%
  filter(model %in% c("addint.f", "addint.i")) %>%
  .$wrap

p3 <- ion_effectplot_df_cov_ep_plots %>%
  filter(model == "int") %>%
  .$wrap


# # # create directory for plots
# dir.create(file.path(paste0(getwd(), "/output/plots"), "effectplots"))

# Saving plots into a single pdf
pdf("output/plots/effectplots/add_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p1, ~ print(.x))
dev.off()

# Saving plots into a single pdf
pdf("output/plots/effectplots/addint_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p2, ~ print(.x))
dev.off()

# Saving plots into a single pdf
pdf("output/plots/effectplots/int_effect_plots_all.pdf",
    onefile = TRUE,  width = 15, height = 9)
walk(p3, ~ print(.x))
dev.off()


```

genomic ranges

```{r}

ion_qtl_granges_example <- qtl_detected_all %>%
  select(seqnames = chr, 
         start = ci.low, 
         end = ci.high,
         phenotype, model) %>%
  arrange(seqnames, start) %>%
  mutate(meta = paste0(phenotype, "-", model)) %>%
  select(-c(phenotype, model)) %>%
  mutate(strand = "*") %>%
  as_granges()

g <- ion_qtl_granges_example

g %>%
  disjoin() %>%
  reduce

g[is.na(
  findOverlaps( 
    g, 
    drop.self = TRUE,
    drop.redundant = T,
    select = "arbitrary"
    ))] %>%
  as_tibble %>%
  mutate(seqnames = as.integer(seqnames))

ion_qtl_granges_example %>%
  reduce %>%
  as_tibble

ion_overlaps <- findOverlaps(
  ion_qtl_granges_example, 
  drop.self = F, 
  drop.redundant = TRUE
  ) 

ion_overlaps

ion_qtl_granges_id <- ion_qtl_granges_example %>%
  as_tibble %>%
  mutate(id = row_number()) %>%
  dplyr::rename(chr = seqnames)
```

Multi-QTL models

```{r}

interval_granges_list_df <- qtl_detected_all %>%
  select(seqnames = chr, 
         start = ci.low, 
         end = ci.high,
         phenotype, model) %>%
  group_by(phenotype, seqnames) %>%
  filter(n() != 1) %>%
  arrange(phenotype) %>%
  mutate(group = cur_group_id())

interval_granges_list <- interval_granges_list_df %>%
  group_split() %>%
  map(.x =., 
      .f = ~ as_granges(.x)) %>%
  GRangesList

reduced_intervals <- interval_granges_list %>%
  GenomicRanges::reduce(.) %>%
  as_tibble 

reduced_intervals_info <- interval_granges_list_df %>%
  left_join(
    reduced_intervals,
    by = c("group", "seqnames")
    ) %>%
  group_by(seqnames, phenotype, group) %>%
  mutate(
    int = grepl("int|WT|HUN", model),
    int = ifelse(sum(int == T) == 0, F, T)
  ) %>%
  select(chr = seqnames, phenotype, group, start.y, end.y, width, 
         int) %>%
  distinct %>%
  ungroup

reduced_intervals_info2 <- qtl_detected_all %>%
  left_join(reduced_intervals_info, by = c("chr", "phenotype")) %>%
  arrange(group) %>%
  mutate(
    pos2 = pmap_lgl(
      .l = .,
      .f = ~ with(
        list(...),
        dplyr::between(pos, start.y, end.y)
        )
      )
    ) %>%
  filter(pos2 == T) %>%
  group_by(phenotype, chr, group) %>%
  select(-model) %>%
  distinct %>%
  filter(lod == max(lod)) %>%
  ungroup %>%
  select(phenotype, chr, pos, int)


qtl_detected_all_reduced <- qtl_detected_all %>%
  anti_join(reduced_intervals_info2, by = c("chr", "phenotype")) %>%
  mutate(
    int = grepl("int|WT|HUN", model),
    int = ifelse(sum(int == T) == 0, F, T)
  ) %>%
  bind_rows(reduced_intervals_info2) %>%
  select(phenotype, chr, pos, int) %>%
  group_by(phenotype, chr)

hun <- data.frame(hun = hun_cov)

qtl_models_all <- qtl_detected_all_reduced %>%
  mutate(pos = round(pos,0)) %>%
  arrange(phenotype, chr) %>%
  group_by(phenotype) %>%
  mutate(id = as.character(row_number()),
         term = ifelse(int == "F",
                       paste("Q", id, sep =""),
                       paste("hun*Q", id, sep = ""))) %>%
  left_join(ion_pheno.col) %>%
  mutate(chr = as.integer(chr)) %>%
  summarise(
    chr = list(chr), 
    pos = list(pos),
    formula = paste("y ~ ", 
                    paste(term, collapse = "+"), 
                    sep = ""),
    pheno.col = first(pheno.col)) %>%
  mutate(qtl_model = pmap(
    .l = .,
    .f = ~ with (
      list(...),
      makeqtl(
        cross = ion_crossobject,
        chr = chr,
        pos = pos
    )))) %>%
    mutate(fit_qtl = pmap(
    .l = .,
    .f = ~ with (
      list(...),
      fitqtl(
        cross = ion_crossobject,
        pheno.col = pheno.col,
        qtl = qtl_model,
        formula = formula, 
        covar = hun
      )
    )))

qtl_full_model_results 
qtl_models_all %>%
  select(phenotype, fit_qtl)  %>%
  mutate(
    formula = map_chr(
      .x = fit_qtl,
      .f = ~ .x  %>% attr(., "formula")),
    resultados = map(
      .x = fit_qtl,
      .f = ~ .x %>% 
        .$result.full %>%
        as_tibble(., rownames = "term"))) %>%
  unnest(resultados) %>%
  select(-fit_qtl) %>%
  filter(term == "Model") %>%
  rename(
    p.Chi2 = `Pvalue(Chi2)`,
    p.F = `Pvalue(F)`
    ) %>%
  mutate(
    s.Chi2 = p.value_to_ast(p.Chi2),
    s.F = p.value_to_ast(p.F),
    term = "Full Model"
    ) %>%
  select(phenotype:`%var`, contains(".Chi2"), contains(".F")) %>%
  mutate(across( c(df:`%var`), ~ round(., 2)),
         across(c(p.Chi2, p.F), 
                ~ formatC(., format = "e", digits = 2)))


                                   


```

